plugins/catalog-backend-module-github/src/providers/GithubOrgEntityProvider.ts

```diff
@@ -224,6 +261,335 @@ export class GithubOrgEntityProvider implements EntityProvider {
+      merge(group, {
+        metadata: {
+          name: slug,
+          annotations: {
+            [`backstage.io/edit-url`]: `${url}/edit`,
+            [`github.com/team-slug`]: `${org}/${slug}`,
+            [ANNOTATION_LOCATION]: `url:${url}`,
+            [ANNOTATION_ORIGIN_LOCATION]: `url:${url}`,
+          },
+        },
+        spec: {
+          profile: {
+            displayName: newTeamName,
+          },
+        },
+      });
+    }
+
+    if (description) {
+      group.metadata.description = description;
+    }
+
+    const { added } = createDeltaOperation(org, [group]);
+    await this.connection.applyMutation({
+      type: 'delta',
+      removed,
+      added,
+    });
+  }
+
+  private async onMemberChangeInOrganization(
+    event: OrganizationMemberAddedEvent | OrganizationMemberRemovedEvent,
+    createDeltaOperation: DeferredEntitiesBuilder,
+  ) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    const userTransformer =
+      this.options.userTransformer || defaultUserTransformer;
+    const { name, avatar_url: avatarUrl, email, login } = event.membership.user;
+    const org = event.organization.login;
+    const { headers } = await this.credentialsProvider.getCredentials({
+      url: this.options.orgUrl,
+    });
+    const client = graphql.defaults({
+      baseUrl: this.options.gitHubConfig.apiBaseUrl,
+      headers,
+    });
+
+    const user = (await userTransformer(
+      {
+        name,
+        avatarUrl,
+        login,
+        email: email || undefined,
+        // we don't have this information in the event, so the refresh will handle that for us
+        organizationVerifiedDomainEmails: [],
+      },
+      {
+        org,
+        client,
+        query: '',
+      },
+    )) as Entity;
+
+    const { added, removed } = createDeltaOperation(org, [user]);
+    await this.connection.applyMutation({
+      type: 'delta',
+      removed,
+      added,
+    });
+  }
+
+  private async onMemberChangeToTeam(
+    event: MembershipEvent,
+    createDeltaOperation: DeferredEntitiesBuilder,
+    updateEntities: EntityUpdateOperation,
+  ) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    // The docs are saying I will receive the slug for the removed event, but the types don't reflect that,
+    // so I will just check to be sure the slug is there
+    // https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#membership
+    if (!('slug' in event.team)) {
+      return;
+    }
+    const teamSlug = event.team.slug;
+    const userSlug = event.member.login;
+
+    const { token } = await this.options.tokenManager!.getToken();
+    const org = event.organization.login;
+    let group = (await this.options.catalogApi!.getEntityByRef(
+      {
+        kind: 'Group',
+        namespace: DEFAULT_NAMESPACE,
+        name: teamSlug,
+      },
+      { token },
+    )) as GroupEntity;
```

Comment:
I'd like to avoid this pattern. There's no guarantee that the shape of the entity is what you expect here, there can be processors that mutate the entities and the catalog model might've been customized. It's also data flowing in the wrong direction, what we're looking for is `github -> providers -> processors -> output entities -> catalog clients`, without any circling back.

Instead of fetching these entities from the catalog, I think it's best to fetch any information that's missing from GitHub instead. So in this case you might fetch the group and all of its members again, or the user, whichever is easiest. Basically, any time you get an event that something has changes, you go back to the source of truth, in this case GitHub, to fill in just enough information that you're able to emit the updated entities.

CC: @pjungermann ^ 👀 

Bit of a separate note here, but I don't think you need to update both the user and group in this case, should be enough to just update one of them and the catalog processing will fill in the relations in both directions.

Reply:
Also not so sure about that as a general rule. E.g., at the `BitbucketCloudEntityProvider` I used the `catalogApi` as well. Different use case though. I needed it to check what exists at catalog in order to create the delta. The events don't contain enough information to do this without.

Reply:
@Rugvip I was working here and when team changes name, I don't have enough information to delete the old relation.

I can use a simple approach and rebuild the whole org (calling the `read` again), or drop support for this event.

The member change is tricky too, because when I emit the delta for the user I need to emit all `memberOf`, so for this we can rebuild the whole org (calling the `read` again), or drop support for this event.

The only drawback I can see  for rebuild th whole org  is for big orgs. We have a hunge chunk to process.
The rate limit will be used for any request, so we can't avoid that.

I will change to call the `read`, but want to get your opnion here too

Reply:
@angeliski for the team name change, I think the old name should be available in the event payload right? at `.changes.name.from`. Or is there some other piece of information that's missing?

For member changes, I'd expect it to be possible to just fetch the entire team and its members again, and apply that as an update?

It'd be best if we can avoid full `.read()` calls except in very rare situations.